src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs

```diff
@@ -328,144 +328,580 @@ private void TrySetUnrecoverableError(SyntaxDiagnosticInfo error)
                 ArrayBuilder<Interpolation>? interpolations,
                 out Range closeQuoteRange)
             {
-                kind = _kind;
-                ScanInterpolatedStringLiteralStart(out openQuoteRange);
-                ScanInterpolatedStringLiteralContents(interpolations);
-                ScanInterpolatedStringLiteralEnd(out closeQuoteRange);
+                // Scan through the open-quote portion of this literal, determining important information the rest of
+                // the scanning needs.
+                var start = _lexer.TextWindow.Position;
+                var succeeded = ScanOpenQuote(out kind, out var startingDollarSignCount, out var startingQuoteCount);
+                Debug.Assert(_lexer.TextWindow.Position != start);
+
+                openQuoteRange = start.._lexer.TextWindow.Position;
+                Debug.Assert(kind != default);
+
+                if (!succeeded)
+                {
+                    // Processing the start of this literal didn't give us enough information to proceed.  Stop now,
+                    // terminating the string to the furthest point we reached.
+                    closeQuoteRange = _lexer.TextWindow.Position.._lexer.TextWindow.Position;
+                    return;
+                }
+
+                ScanInterpolatedStringLiteralContents(kind, startingDollarSignCount, startingQuoteCount, interpolations);
+                ScanInterpolatedStringLiteralEnd(kind, startingQuoteCount, out closeQuoteRange);
             }
 
-            private void ScanInterpolatedStringLiteralStart(out Range openQuoteRange)
+            /// <param name="startingDollarSignCount">
+            /// Number of '$' characters this interpolated string started with.  We'll need to see that many '{' in a
+            /// row to start an interpolation.  Any less and we'll treat that as just text.  Note if this count is '1'
+            /// then this is a normal (non-raw) interpolation and `{{` is treated as an escape.
+            /// </param>
+            /// <param name="startingQuoteCount">Number of '"' characters this interpolated string started with.</param>
+            /// <returns><see langword="true"/> if we successfully processed the open quote range and can proceed to the
+            /// rest of the literal. <see langword="false"/> if we were not successful and should stop
+            /// processing.</returns>
+            private bool ScanOpenQuote(
+                out InterpolatedStringKind kind,
+                out int startingDollarSignCount,
```

Comment:
Consider using `uint` for counts since cannot be negative

Reply:
interesting.  do you guys have precedent for doing that internally?

Reply:
ok.  so trying this leads to a bunch of fallout that is unpleasant (like interacting with other very common apis that are int based).  I'd prefer to not do this if that's ok?

Reply:
Okay. Not worth it then